symengine/series_generic.cpp

```diff
@@ -2,183 +2,239 @@
-                 [&](decltype(v[0]) i) { return i < prec_; });
-    poly_ = UnivariateIntPolynomial::create(var_, vtrunc);
+RCP<const Basic> UnivariateSeries::get_coeff(int deg) const
+{
+    if (p_.get_univariate_poly()->get_dict().count(deg) == 0)
+        return zero;
+    else
+        return p_.get_univariate_poly()->get_dict().at(deg).get_basic();
 }
 
-bool UnivariateSeries::is_canonical(const UnivariateIntPolynomial &poly,
-                                    const unsigned int &prec) const
+UnivariateExprPolynomial UnivariateSeries::var(const std::string &s)
 {
-    return true;
+    return UnivariateExprPolynomial(UnivariatePolynomial::create(symbol(s),
+                                                                 {0, 1}));
 }
 
-std::size_t UnivariateSeries::__hash__() const
+Expression UnivariateSeries::convert(const Basic &x)
 {
-    return poly_->hash() + std::size_t(prec_ * 84728863L);
+    return Expression(x.rcp_from_this());
 }
 
-bool UnivariateSeries::__eq__(const Basic &other) const
+int UnivariateSeries::ldegree(const UnivariateExprPolynomial &s)
 {
-    if (not is_a<UnivariateSeries>(other))
-        return false;
-    const UnivariateSeries &o = static_cast<const UnivariateSeries &>(other);
-    return operator==(o);
+    return s.get_univariate_poly()->get_dict().begin()->first;
 }
 
-int UnivariateSeries::compare(const Basic &other) const
+UnivariateExprPolynomial
+UnivariateSeries::mul(const UnivariateExprPolynomial &a,
+                      const UnivariateExprPolynomial &b, unsigned prec)
 {
-    if (not is_a<UnivariateSeries>(other))
-        throw std::domain_error("cannot compare with UnivariateSeries");
-    const UnivariateSeries &o = static_cast<const UnivariateSeries &>(other);
-    return poly_->compare(*o.poly_);
+    map_int_Expr p;
+    for (auto &it1 : a.get_univariate_poly()->get_dict()) {
+        for (auto &it2 : b.get_univariate_poly()->get_dict()) {
+            int exp = it1.first + it2.first;
+            if (exp < (int)prec) {
+                p[exp] += it1.second * it2.second;
+            } else {
+                break;
+            }
+        }
+    }
+    if (a.get_univariate_poly()->get_var()->get_name() == "")
+        return UnivariateExprPolynomial(UnivariatePolynomial::from_dict(b.get_univariate_poly()->get_var(), std::move(p)));
+    else
+        return UnivariateExprPolynomial(UnivariatePolynomial::from_dict(a.get_univariate_poly()->get_var(), std::move(p)));
 }
 
-std::string UnivariateSeries::__str__() const
+UnivariateExprPolynomial
+UnivariateSeries::pow(const UnivariateExprPolynomial &base, int exp,
+                      unsigned prec)
 {
-    std::ostringstream o;
-    bool first = true;
-    for (const auto &it : poly_->get_dict()) {
-        if (it.second == 0)
-            continue;
-        if (first) {
-            if (it.second < 0)
-                o << "-";
+    if (exp < 0)
+        return UnivariateSeries::pow(UnivariateExprPolynomial::inverse(base), -exp, prec);
```

Comment:
For this, you only need to check that `base` has a dictionary of size 1. Otherwise throw an error. When dictionary size is 1, it is easy to implement.